daemon/libnetwork/drivers/bridge/bridge_linux.go

```diff
@@ -1585,15 +1595,28 @@ func (d *driver) ProgramExternalConnectivity(ctx context.Context, nid, eid strin
 //
 // ep.portMapping is updated when bindings are removed.
 func (ep *bridgeEndpoint) trimPortBindings(ctx context.Context, n *bridgeNetwork, pbmReq portBindingMode) (func() []portmapperapi.PortBinding, error) {
-	// If the endpoint is the gateway for IPv4 and IPv6, there's nothing to drop.
-	if pbmReq.ipv4 && pbmReq.ipv6 {
+	// If the network is IPv4-only, this endpoint is the container's IPv6 gateway,
+	// and IPv4 addresses are published to the host - when it's acting as the IPv6
+	// gateway, bindings may be set up to proxy from host IPv6 to endpoint IPv4.
+	proxy6To4 := ep.addr != nil && ep.addrv6 == nil && !n.gwMode(firewaller.IPv4).routed()
+
+	// Drop IPv4 bindings if this endpoint is not the IPv4 gateway, unless the
+	// network is "routed" (routed bindings get dropped unconditionally by Leave).
+	drop4 := !pbmReq.ipv4 && !n.gwMode(firewaller.IPv4).routed()
+
+	// Drop IPv6 bindings if this endpoint is not the IPv6 gateway, unless the IPv6
+	// network is "routed" (routed bindings get dropped unconditionally by Leave),
+	// or there may be mappings proxying between host IPv6 and container IPv4.
+	drop6 := !pbmReq.ipv6 && (!n.gwMode(firewaller.IPv6).routed() || proxy6To4)
```

Comment:
Based on our Slack discussion, I think you don't need to check `proxy6To4` in `drop6`, because `pbmReq.ipv6` is set to true when the current endpoint is the IPv6 gateway, or when the container has no IPv6 gateway at all and the current endpoint is the IPv4 gateway.

https://github.com/moby/moby/blob/e1281f09fceec4aab518267c319a7bd4c79cf3c5/daemon/libnetwork/drivers/bridge/bridge_linux.go#L1537-L1542

So, you can probably replace `drop6` with:

```
drop6 := !pbmReq.ipv6 && !n.gwMode(firewaller.IPv6).routed()
```

Also, I think it's quite confusing to say `Drop IPv6 bindings if this endpoint is not the IPv6 gateway` while `pbmReq.ipv6` is set true when there's no IPv6 gateway.